unable to create resv from job if flatuid=False AND reject altering of create_resv_from_job if the job is already running

prakashcv13 opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Describe Bug or Feature
<!--- Describe the problem, ideally from the customer's viewpoint  -->
When flatuid is set to False, create_resv_from_job does not work.

#### Describe Your Change
<!--- Say how you fixed the problem.  Please describe your code changes in detail for reviewer -->
**Change 1)** In copy_params_from_job(), the user was being set as below -
snprintf(buf, 255, "%s@%s", pjob->ji_wattr[(int)JOB_ATR_job_owner].at_val.at_str,
pjob->ji_wattr[(int)JOB_ATR_submit_host].at_val.at_str);

Now, added a check to see if '@' is already present and copy the owner accordingly.

**Change 2)**
Also, @arungrover reported a crash with the below trace -
(gdb) bt
#0  0x00007f4e01c08ada in __strcmp_sse42 () from /usr/lib64/libc.so.6
#1  0x0000000000458915 in find_resv_by_quename (quename=quename@entry=0x1319088 "") at ../../../src/server/queue_func.c:355
#2  0x0000000000469ca9 in req_quejob (preq=0x1318cf0) at ../../../src/server/req_quejob.c:677
#3  0x00000000004bd0bc in process_socket (sock=sock@entry=15) at ../../../../src/lib/Libnet/net_server.c:512
#4  0x00000000004bd282 in wait_request (waittime=waittime@entry=2, priority_context=<optimized out>) at ../../../../src/lib/Libnet/net_server.c:628
#5  0x000000000042b55b in main (argc=<optimized out>, argv=<optimized out>) at ../../../src/server/pbsd_main.c:1833
(gdb) fr 2
#2  0x0000000000469ca9 in req_quejob (preq=0x1318cf0) at ../../../src/server/req_quejob.c:677
677                                         presv = find_resv_by_quename(qname);
(gdb) p qname
$1 = 0x1319088 ""
(gdb)

Added a NULL check to prevent the crash and also moved this check inside the if where it is relevant.

**Change 3)**
Rejecting attempt to alter create_resv_from_job if the job is running.

#### Link to Design Doc
<!--- If there is a design, link to it here: **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)** -->


#### Attach Test and Valgrind Logs/Output
<!--- Please attach your test log output from running the test you added (or from existing tests that cover your changes) -->
<!--- Don't forget to run Valgrind if appropriate and attach the resulting logs -->

<!--- Pull Request Guidelines: [Pull Request Guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/1187348483/Pull+Request+Guidelines) -->
[before-24606.txt](https://github.com/PBSPro/pbspro/files/4419797/before-24606.txt)
[after-24606.txt](https://github.com/PBSPro/pbspro/files/4419798/after-24606.txt)

bhroam commented on test/tests/functional/pbs_user_reliability.py:
```diff
@@ -109,11 +109,16 @@ def test_create_resv_from_job_using_qsub(self):
 
+        self.server.log_match("Reject reply code=15095", starttime=now,
+                              interval=2, max_attempts=20, existence=False)
+
```
This is a forced 40s wait in the test.  Is this what you want?  Maybe max_attempts can be 10?

bhroam commented on src/server/req_quejob.c:
```diff
@@ -671,12 +671,16 @@ req_quejob(struct batch_request *preq)
-		presv = find_resv_by_quename(qname);
+		if (index == JOB_ATR_create_resv_from_job) {
+			if (qname != NULL && *qname != '\0') {
+				presv = find_resv_by_quename(qname);
```
If this is the only time you use presv, move the declaration to this block.  Also, you added a initialization to NULL, I'm not sure you need it.

prakashcv13 commented:
hey @bhroam - thank you for reviewing the code. I have addressed your review comments. PFA the test logs. 
[ptl-24606-2.txt](https://github.com/PBSPro/pbspro/files/4424363/ptl-24606-2.txt)

riyazhakki commented:
@prakashcv13 , Just a doubt, when flatuid is set to False, users who have no account at the server host can create_resv_from_job from his execution host?

prakashcv13 commented:
@riyazhakki - thank you for reviewing the PR - 
flatuid just says whether, for each user, the username at the submission host must be the same as the one at the server host.
when it is set to False - if the user is able to submit a job - he should be able to create a reservation. That is what was not working - it is fixed through this PR.

prakashcv13 commented:
hey @bhroam and @riyazhakki - I have made some more changes - 
changing create_resv_from_job=1 after the job starts running, does not have any effect.
With the new changes, we do not allow it to be changed.
[after-qalter.txt](https://github.com/PBSPro/pbspro/files/4426361/after-qalter.txt)
[before-qalter.txt](https://github.com/PBSPro/pbspro/files/4426362/before-qalter.txt)

riyazhakki commented on test/tests/functional/pbs_user_reliability.py:
```diff
@@ -225,3 +230,29 @@ def test_create_resv_by_other_user(self):
+
+    def test_set_attr_when_job_running(self):
+        """
+        This test confirms that create_resv_from_job=1 is not allowed to be
```
```suggestion
        This test confirms that create_resv_from_job is not allowed to be
```

bhroam commented on test/tests/functional/pbs_user_reliability.py:
```diff
@@ -238,3 +238,21 @@ def test_flatuid_false(self):
+            msg += "create_resv_from_job"
+            self.assertTrue(msg in e.msg[0])
+        else:
+            self.fail("attribute allowed to be modified")
```
Use assertRaises (and assertIn):
```suggestion
           with self.assertRaises(PbsAlterError, msg="attribute allowed to be modified") as e:
               msg = "..."
               self.assertIn(msg, e.msg[0])
```
